cmd/state_delete.go

```diff
@@ -0,0 +1,66 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi/pkg/resource"
+	"github.com/pulumi/pulumi/pkg/resource/deploy"
+	"github.com/pulumi/pulumi/pkg/resource/edit"
+	"github.com/pulumi/pulumi/pkg/util/cmdutil"
+	"github.com/spf13/cobra"
+)
+
+func newStateDeleteCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "delete",
+		Short: "Deletes a resource from a stack's state",
+		Long: `Deletes a resource from a stack's state
+		
+This command deletes a resource from a stack's state, as long as it is safe to do so. Resources can't be deleted if
+there exist other resources that depend on it or are parented to it.`,
```

Comment:
or are protected?

Reply:
For a raw state-editing command, I figured that it was totally fine to delete protected resources. I rationalized this on two fronts:

1. The engine's notion of "protected" is that "a provider's `Delete` will never be called on this resource", which is still true if a resource is deleted via `pulumi state delete`
2. In my mind `pulumi state` is a C++ kinda tool, which is to say that it does exactly what you tell it to do and doesn't get in your way. You ran `pulumi state delete`? We're going to delete this thing unless we absolutely can't do it.

Reply:
i feel like i don't like that.  it feels like a surprising behavior that could upset people.  "i assumed this could only delete resources safe to delete!"

i'm curious what others think though...

Reply:
I would definitely prefer a `--force` option (or `--unprotected`) here.

Reply:
OK. Done in the most recent commit.